### PR TITLE
Project Todo - Sandra Wallén

### PR DESCRIPTION
**Pull request by sandra-wallen**
**Link to my deployed project:**
https://sandras-todos.netlify.app

**My reflections on how this week's project turned out:**
I'm overall very pleased with how the project turned out. I set up a "data base" with a REST API and put a lot of time on this. The app sends GET, POST, DELETE and PUT requests to the REST API. The PUT request is sent when an item is toggled complete or the item text is updated.

When there are no todos to be displayed an empty state component is mounted and x/x tasks + the check all/uncheck all button component is only mounted when there are items in the data base.

I had many ideas for this project but I didn't have the time to implement them all. See the readme file for a further explanations about me project and the things I would want to implement/fix next

**Things I'd like to have clarified or explained in more detail:**
Since it's always some response time for the API I have also implemented local updates to the store that will show until the response. There's still some glitches in this functionality however, sometimes when an item is deleted the local state updates and filters out that item but then the refetch is executed too quickly, resulting in the deleted item still being in the response and now showing in the list again. I set a setTimeout on the refetches but that didn't do the trick, I will have to investigate this further. Another thing that you should have in mind when using my application is that the restdb didn't provide any documentation on how to do a batch request, or if it's even possible. Therefor the checkAll and uncheckAll functions are just updating locally and will not send any requests to the API.

I have not made the local state update while waiting for the response from the POST request because I didn't know how to think when sending this to the intialState locally, in the object there is an idgenerator and a moment function and if I would just recreate this when updating the state locally it would not generate the same values as the real request sent to the API.

Also, I struggled alot with styling the datepicker initially, and after a while a found that I could overwrite the provided stylesheet for the component but I didn't know how to do this as a styled component, the datepicker is the only element that is styled with vanilla css.

### CODE REVIEW

File: code/src/components/CheckAllBtn.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    }
+  }
+
+  return(
+    <>
+      {store.items.length > 0 && (<Btn 
+        type="button" 
+        onClick={onBtnClick}
+      >
+        {store.allChecked ? "UNCHECK ALL" : "CHECK ALL"}
```
Comment (annalenksjo) [inline code/src/components/CheckAllBtn.js:43]: Nice implementation with the change of button text!

File: code/src/components/Header.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  justify-content: space-evenly;
+`
+
+const ListHeading = styled.h2`
+  margin: 0;
+`
+
+const Date = styled.p`
+  
+`
```
Comment (annalenksjo) [inline code/src/components/Header.js:49]: Is this intentionally blank to create a semantic tag?

File: code/src/components/Header.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+const Header = () => {
+  const today = moment().format('MMMM Do')
+  const weekday = moment().format('dddd')
```
Comment (annalenksjo) [inline code/src/components/Header.js:53]: Nice that you did the moments to constants, makes it easier to follow!

File: code/src/components/TodoItem.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  font-size: 20px;
+`
+
+const Date = styled.span`
+  font-size: 12px;
+`
+
+const TodoItem = ({ id, text, complete, editMode, dueDate, description, setDescription }) => {
+  const dispatch = useDispatch()
+  const taskDueDate = moment(dueDate).format('MMM Do')
```
Comment (annalenksjo) [inline code/src/components/TodoItem.js:66]: 👍

File: code/src/components/TodoList.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+          text={item.text}
+          complete={item.complete}
+          editMode={item.editMode}
+          dueDate={item.dueDate}
+          description={description}
+          setDescription={setDescription}
+        />
+      ))}
+      {items.length < 1 && (
+        <EmptyState />
```
Comment (annalenksjo) [inline code/src/components/TodoList.js:61]: Nice! I liked your empty state pic too!

File: code/src/index.css [hunk 6]
```diff
@@ -11,3 +11,45 @@ code {
...
+  color: #8B98F9;
+}
+
+.testing .bx--date-picker__icon {
+  fill: #8B98F9;
+  padding-left: 0;
+  margin-left: auto;
+}
+
+.bx--date-picker.bx--date-picker--single .bx--date-picker__input {
```
Comment (annalenksjo) [inline code/src/index.css:36]: wow complicated classnames :P

**Review by annalenksjo — APPROVED** [review]
Awesome project, nice design! It was cool that you really pushed the limits and I know you put in a lot more than half-time. The readme was extensive (good job) and it was interesting to follow your process and struggles. 

As you mention there are some remaining bugs. So I guess its kind of a still ongoing project :)

You really did a good job in dividing into (many) components as well. Well structured, good naming. 

Maybe add some other styling to the tasks that are checked, to more easily distinguish them from the ones that are left to do. Also, it would be smooth to add a task with the enter-key as well! 

But all in all great job, a high ambition I applaud you for that!

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
Please check your netlify link because it is returning 404 Error 😭

File: code/src/components/CustomCheckbox.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  ${Input}:checked + &::before {
+    display: block;
+  }
+  ${Input}:hover + &::before{
+    background-color: #D6A1A2;
+  }
+  ${Input}:hover + & {
+    background-color: #f6e6e4;
+  }
+`
```
Comment (KarolCislowski) [inline code/src/components/CustomCheckbox.js:57]: Impressive 😮

File: code/src/components/CustomDatePicker.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        <DatePickerInput
+          hideLabel
+          id="date-picker-default-id"
+          className="testing-input"
+          size="md"
+          placeholder="due date"
+          labelText="Due Date"
+          type="text"
+        />
+      </DatePicker>
```
Comment (KarolCislowski) [inline code/src/components/CustomDatePicker.js:31]: Really nice 🥇

File: code/src/reducers/tasks.js [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+      const addFetchedData = action.payload.map(item => {
+        return { _id: item._id, id: item.id, text: item.text, complete: item.complete, created: item.created, editMode: item.editMode, dueDate: item.dueDate }
+      })
+
+      const sortedData = addFetchedData.sort((a, b) => {
+        return new Date(a.dueDate) - new Date(b.dueDate)
+      })
+
+      store.items = sortedData
+    },
```
Comment (KarolCislowski) [inline code/src/reducers/tasks.js:46]: 😮

**Review by KarolCislowski — APPROVED** [review]
Great APP! I can see you mastered the process of building good, reusable components, which makes your code super clean and easy to maintain 🦸🏻‍♀️

**Pull request closed without merging**